API usage examples

theri opened the pull request:
This PR provides usage examples of our abstract interface, to address #247.

The examples show how an application might use Listen(), Initiate(), and Rendezvous(), send a Message, and receive a Message.
There is some redundancy between these examples, and I introduces some subtle differences to make them plausible, e.g., the "server" provides its identity, and the "client" verifies the other endpoint's identity using a trust callback.

Overall the section became rather long. Should we instead just stick to one example, e.g., Initiate()? (Or Rendezvous() because it covers the most aspects? Or Initiate() but with all the aspects from Rendezvous()?)

Otherwise I'm particularly interested in feedback on the Rendezvous() example -- I think it's correct wrt the draft, but is there a "more plausible" port number, and should anything else be set there that is not in the Initiate() and Listen() examples?

tfpauly commented on draft-ietf-taps-interface.md:
```diff
@@ -320,6 +483,7 @@ following recommendations:
   because transport protocols implementing a given Property are not available on
   the platform.
 
+
```
Random extra line? =)

theri replied:
Oops, yea, that happened when I moved this part around on my branch...

tfpauly commented on draft-ietf-taps-interface.md:
```diff
@@ -258,6 +258,169 @@ Errors and other notifications also happen asynchronously on the Connection.
+## Usage Examples
+
+The following usage examples illustrate how an application might use a
+Transport Services Interface to
```
Editorial nit, feels like this should be structured like:

```
how an application might use a Transport Services Interface to:

- Act as a server, by listening for incoming connections, receiving requests, and sending responses. See {{server-example}}.
```
etc. The main point being we should have a `:` on the dangling sentence, which is filled out by the bullets.

theri replied:
Agree, will change.

tfpauly commented on draft-ietf-taps-interface.md:
```diff
@@ -258,6 +258,169 @@ Errors and other notifications also happen asynchronously on the Connection.
+TransportProperties := NewTransportProperties()
+TransportProperties.Require(reliability)
+TransportProperties.Require(preserve-order)
+TransportProperties.Require(preserve-msg-boundaries)
```
I'd actually argue that we should not show setting these values. If we do believe that these are the defaults:

```
The recommended default is to Require Reliable Data Transfer.
```

Then why would the application need to set these explicitly? That just adds to boiler-plate code bulk.

theri replied:
True for reliability and preserve order -- I wanted to show them explicitly, but I can also do that in a comment, and that's probably better.
For preserve-msg-boundaries, the default is Prefer, so I do think I need to set it here, otherwise I think the example will be somewhat ambiguous.

tfpauly replied:
Hm, okay. I think keeping one example of a changed property is good, so this one is fine. May want to change it in the future, but we can see.

britram replied:
I _think_ we're leaning toward underscore-, as opposed to dash-separation for property names? in any case I don't think this has been decided yet so can leave it for now...

tfpauly commented on draft-ietf-taps-interface.md:
```diff
@@ -258,6 +258,169 @@ Errors and other notifications also happen asynchronously on the Connection.
+
+~~~
+LocalSpecifier := NewLocalEndpoint()
+LocalSpecifier.WithInterface("en0")
```
I don't like ever recommending using an interface by name; people often make incorrect assumptions about which interface is named which.

theri replied:
Agreed, hardcoding these in an application might be a bad idea.
I took this example from the Transport Properties Section, so if we decide it's bad, maybe we should change it there as well.

What about: LocalSpecifier.WithInterface("any")? Or is there a better way to specify an interface to listen on than with interface names?

tfpauly commented on draft-ietf-taps-interface.md:
```diff
@@ -258,6 +258,169 @@ Errors and other notifications also happen asynchronously on the Connection.
+
+// Only receive complete messages
+minIncompleteLength = Infinite
+maxLength = Infinite
```
Again, aren't these the defaults?

theri replied:
Yes, they are.
I wanted to show them explicitly, so readers are not confused when they read the parameters of the Receive call. But again, I can do that in a comment and that's probably better.

theri commented:
Thanks for the review, I addressed the comments.
Maybe the default for Preserve Message Boundaries should be Require as well, if we receive only complete Messages by default...

tfpauly commented on draft-ietf-taps-interface.md:
```diff
@@ -258,6 +258,163 @@ Errors and other notifications also happen asynchronously on the Connection.
+
+// Only receive complete messages:
+// minIncompleteLength and maxLength are infinite by default
+Connection.Receive(minIncompleteLength, maxLength)
```
Would the defaults not make this collapse down to Connection.Receive()?

theri replied:
You're right, I overlooked the "?".

tfpauly commented:
I'm good with these changes. Regarding how we show the pseudo-code, I'm not totally sure about the notation (how rigorous are we on that?), so someone else should validate it makes sense.

britram commented on draft-ietf-taps-interface.md:
```diff
@@ -258,6 +258,160 @@ Errors and other notifications also happen asynchronously on the Connection.
+
+Preconnection.Listen()
+
+Preconnection -> ConnectionReceived<Connection>
```
this means that "the following lines will happen after this event triggers"? 

We should probably add a paragraph or two to the notation section for notations we want to use in examples. As above, see no reason to block merging on this though.,
